### ISSUE TITLE
Questions on the loss function of distance 

### ISSUE THREAD

**Issue by BirkhoffG** [interpretml/DiCE#44]
I have one question about the loss function in measuring the proximity between generated counterexamples `c` and input instance `x`. From your [paper](http://arxiv.org/abs/1905.07697), formula (5) and (6) illustrates the distance loss function. However, I understand (6) is not a differentiable function. How do you cope with this problem?

I also check your code in `dice_pytorch.py`, and it seems like you only measure the distance by calculating the L_1 distance. 

https://github.com/interpretml/DiCE/blob/a772c8d4fcd88d1cab7f2e02b0bcc045dc0e2eab/dice_ml/explainer_interfaces/dice_pytorch.py#L208-L217

Thank you for your time. Looking forward to your reply.

**Comment by raam93** [comment]
Hi, we explain how we implement formula (6) in practice in the paragraph (with heading **Relative scale of features**) just below the formula. I quote the following excerpt for your reference:

> For categorical features, we convert each feature using one-hot encoding and consider it as a continuous variable between 0 and 1. Also, to enforce the one-hot encoding in the learned counterfactuals, we add a regularization term with high penalty for each categorical feature to force its values for different levels to sum to 1. At the end of the optimization, we pick the level with maximum value for each categorical feature.

Does this answer your question?

**Comment by BirkhoffG** [comment]
Thanks for your reply.

Yes, that is what I am confused about. It makes sense that you treat the categorical feature as a continuous feature with one-hot encoding. It might be better to explicitly point out that paragraph is a practical consideration in implementing (6).

However, I am still confused about how to enforce the one-hot encoding in the learned counterfactuals. 

> Also, to enforce the one-hot encoding in the learned counterfactuals, we add a regularization term with high penalty for each categorical feature to force its values for different levels to sum to 1. At the end of the optimization, we pick the level with maximum value for each categorical feature.

Could you help me to understand how do you ensure each categorical feature to sum to 1 by adding a regularization term? Why not adding a softmax layer to ensure each categorical feature to sum to 1?

**Comment by raam93** [comment]
We add L2 regularization to ensure the sum to 1 (https://github.com/interpretml/DiCE/blob/master/dice_ml/explainer_interfaces/dice_pytorch.py#L258). During each optimization step, we do a bunch of tasks such as freezing some gradients (if some features are not allowed to change). So are you suggesting to add a softmax layer after each optimization step or to include the softmax function as a loss component? I think both can be done and it's a great suggestion! Feel free to raise a PR and it will be interesting to compare the results from the two approaches.

**Comment by BirkhoffG** [comment]
Thanks for the reply! I will certainly look into the difference between L2 regularization and softmax.
